openshift/catalog-sync/pkg/check/catalog.go

```diff
@@ -0,0 +1,143 @@
+package check
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+func AllCatalogChecks() []CatalogCheck {
+	return []CatalogCheck{
+		BundleHasValidMetadataStructure(),
+		ChannelHeadsHaveCSVMetadata(),
+	}
+}
+func BundleHasValidMetadataStructure() CatalogCheck {
+	return CatalogCheck{
+		Name: "CheckBundleHasValidMetadataStructure",
+		Fn: func(ctx context.Context, fbcFS fs.FS) error {
+			const configsDir = "configs"
+
+			entries, err := fs.ReadDir(fbcFS, configsDir)
+			if err != nil {
+				return fmt.Errorf("read configs dir: %w", err)
+			}
+
+			var errs []error
+
+			for _, entry := range entries {
+				if !entry.IsDir() {
+					continue
+				}
+				pkgPath := filepath.Join(configsDir, entry.Name())
+
+				subFS, err := fs.Sub(fbcFS, pkgPath)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("sub FS for %q: %w", pkgPath, err))
+					continue
+				}
+
+				cfg, err := declcfg.LoadFS(ctx, subFS)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("load declcfg from %q: %w", pkgPath, err))
+					continue
+				}
+
+				for _, bundle := range cfg.Bundles {
+					var hasCSV, hasObject bool
+					for _, prop := range bundle.Properties {
+						if prop.Type == "olm.csv.metadata" {
+							hasCSV = true
+						}
+						if prop.Type == "olm.bundle.object" {
+							hasObject = true
+						}
+					}
+
+					if (hasCSV && hasObject) || (!hasCSV && !hasObject) {
+						errs = append(errs, fmt.Errorf("bundle %q in package %q has both or neither of olm.bundle.object / olm.csv.metadata", bundle.Name, bundle.Package))
+					}
+				}
```

Comment:
This isn't quite right for what we care about.

This set of checks is for 4.20. In 4.20, we expect:
- every channel head to have `olm.csv.metadata`.
- no bundle to have `olm.bundle.object`.

With my above comment in mind, I would suggest:
1. Getting a `declcfg.DeclarativeConfig` for the entire catalog
2. Converting that to a model (at which point you'll have a map of package names)
3. For each package:
   a. For each channel, get the channel head and check that `olm.csv.metadata` is present
   b. For each bundle, check that `olm.bundle.object` is NOT present.

Reply:
.... I scroll, and now I see the next check is close to what I'm suggesting. We don't need both checks. I think we can just have the one check for "expected operator metadata" that checks what I describe above.

Reply:
Done